File: demo-apps/demoapp002-python-pytest/tests/screenplay/fixtures/grid_fixtures.py
from __future__ import annotations

from app_src import Puzzle, SudokuSolver
from app_src.constants import BLOCK_SIZE, EMPTY_CELL, GRID_SIZE


def _digits_except(missing_digit: int) -> list[int]:
    return [digit for digit in range(1, GRID_SIZE + 1) if digit != missing_digit]


def setup_almost_complete_column(solver: SudokuSolver, col: int, missing_digit: int) -> None:
    values = _digits_except(missing_digit)
    for index, value in enumerate(values):
        solver.grid[index + 1][col] = value
    solver.grid[0][col] = EMPTY_CELL


def setup_almost_complete_block(
    solver: SudokuSolver, block_row: int, block_col: int, missing_digit: int
) -> None:
    values = _digits_except(missing_digit)
    index = 0
    for row in range(block_row * BLOCK_SIZE, (block_row + 1) * BLOCK_SIZE):
        for col in range(block_col * BLOCK_SIZE, (block_col + 1) * BLOCK_SIZE):
            solver.grid[row][col] = values[index] if index < len(values) else EMPTY_CELL
            index += 1


def setup_multiple_empties(solver: SudokuSolver) -> None:
    for row in range(GRID_SIZE):
        for col in range(GRID_SIZE):
            solver.grid[row][col] = ((row * 3 + col + 1) % 9) + 1 if (row + col) % 2 == 0 else EMPTY_CELL


def setup_row_missing_digit(solver: SudokuSolver, row_index: int, target: int) -> None:
    digits = _digits_except(target)
    index = 0
    for col in range(GRID_SIZE):
        if col == 4:
            solver.grid[row_index][col] = EMPTY_CELL
        else:
            solver.grid[row_index][col] = digits[index]
            index += 1


def setup_column_missing_digit(solver: SudokuSolver, col_index: int, target: int) -> None:
    digits = _digits_except(target)
    index = 0
    for row in range(GRID_SIZE):
        if row == 4:
            solver.grid[row][col_index] = EMPTY_CELL
        else:
            solver.grid[row][col_index] = digits[index]
            index += 1


def setup_block_four_empties(solver: SudokuSolver) -> None:
    solver.grid[0][0] = 1
    solver.grid[1][0] = 2
    solver.grid[2][0] = 3
    solver.grid[2][1] = 6
    solver.grid[2][2] = 7


def setup_hidden_single_in_block(solver: SudokuSolver, target: int) -> None:
    solver.grid[0][5] = target
    solver.grid[3][1] = target


def setup_digit_in_row(solver: SudokuSolver, row_index: int, digit: int) -> None:
    solver.grid[row_index][5] = digit


def clear_cell(solver: SudokuSolver, row: int, col: int) -> None:
    solver.grid[row][col] = EMPTY_CELL


def setup_three_candidates(solver: SudokuSolver) -> None:
    solver.grid[0][1] = 1
    solver.grid[0][2] = 3
    solver.grid[0][3] = 4
    solver.grid[1][0] = 6
    solver.grid[2][0] = 7
    solver.grid[4][0] = 9


def setup_three_naked_singles(solver: SudokuSolver) -> None:
    solution = [
        [5, 3, 4, 6, 7, 8, 9, 1, 2],
        [6, 7, 2, 1, 9, 5, 3, 4, 8],
        [1, 9, 8, 3, 4, 2, 5, 6, 7],
        [8, 5, 9, 7, 6, 1, 4, 2, 3],
        [4, 2, 6, 8, 5, 3, 7, 9, 1],
        [7, 1, 3, 9, 2, 4, 8, 5, 6],
        [9, 6, 1, 5, 3, 7, 2, 8, 4],
        [2, 8, 7, 4, 1, 9, 6, 3, 5],
        [3, 4, 5, 2, 8, 6, 1, 7, 9],
    ]
    for row in range(GRID_SIZE):
        for col in range(GRID_SIZE):
            solver.grid[row][col] = solution[row][col]
    solver.grid[0][0] = EMPTY_CELL
    solver.grid[4][4] = EMPTY_CELL
    solver.grid[8][8] = EMPTY_CELL


def setup_named_grid_state(solver: SudokuSolver, grid_state: str) -> None:
    if grid_state == "has5InSameRow":
        solver.grid[0][5] = 5
    elif grid_state == "has3InSameCol":
        solver.grid[5][0] = 3
    elif grid_state == "has7InSameBlock":
        solver.grid[0][0] = 7
    elif grid_state == "has1InRowAndCol":
        solver.grid[8][0] = 1
        solver.grid[0][8] = 1
    elif grid_state == "fullyConstrained":
        solver.grid[3][0] = 8


def setup_with_duplicate_in_row(solver: SudokuSolver, row_index: int, value: int) -> None:
    solver.grid[row_index][0] = value
    solver.grid[row_index][1] = value


def add_values_to_row(solver: SudokuSolver, row: int, exclude_col: int, values: list[int]) -> None:
    placed = 0
    for col in range(GRID_SIZE):
        if placed >= len(values):
            break
        if col != exclude_col and solver.grid[row][col] == EMPTY_CELL:
            solver.grid[row][col] = values[placed]
            placed += 1


def add_values_to_column(solver: SudokuSolver, col: int, exclude_row: int, values: list[int]) -> None:
    placed = 0
    for row in range(GRID_SIZE):
        if placed >= len(values):
            break
        if row != exclude_row and solver.grid[row][col] == EMPTY_CELL:
            solver.grid[row][col] = values[placed]
            placed += 1


def add_values_to_block(
    solver: SudokuSolver,
    target_row: int,
    target_col: int,
    exclude_row: int,
    exclude_col: int,
    values: list[int],
) -> None:
    block_start_row = (target_row // BLOCK_SIZE) * BLOCK_SIZE
    block_start_col = (target_col // BLOCK_SIZE) * BLOCK_SIZE
    placed = 0
    for row in range(block_start_row, block_start_row + BLOCK_SIZE):
        for col in range(block_start_col, block_start_col + BLOCK_SIZE):
            if placed >= len(values):
                return
            if (row != exclude_row or col != exclude_col) and solver.grid[row][col] == EMPTY_CELL:
                solver.grid[row][col] = values[placed]
                placed += 1


def create_solvers_from_puzzles(count: int, puzzles: list[Puzzle]) -> list[SudokuSolver]:
    return [SudokuSolver(puzzle.name, puzzle.grid) for puzzle in puzzles[:count]]

File: demo-apps/demoapp003-csharp-specflow/tests/screenplay/fixtures/GridFixtures.cs
using DemoApp003.Sudoku;

namespace DemoApp003.Specs.Screenplay.Fixtures;

public static class GridFixtures
{
    public static void SetupAlmostCompleteColumn(SudokuSolver solver, int col, int missingDigit)
    {
        var values = DigitsExcept(missingDigit);
        for (var index = 0; index < values.Length; index++)
        {
            solver.Grid[index + 1][col] = values[index];
        }

        solver.Grid[0][col] = Constants.EmptyCell;
    }

    public static void SetupAlmostCompleteBlock(SudokuSolver solver, int blockRow, int blockCol, int missingDigit)
    {
        var values = DigitsExcept(missingDigit);
        var index = 0;
        for (var row = blockRow * Constants.BlockSize; row < (blockRow + 1) * Constants.BlockSize; row++)
        {
            for (var col = blockCol * Constants.BlockSize; col < (blockCol + 1) * Constants.BlockSize; col++)
            {
                solver.Grid[row][col] = index < values.Length ? values[index] : Constants.EmptyCell;
                index++;
            }
        }
    }

    public static void SetupMultipleEmpties(SudokuSolver solver)
    {
        for (var row = 0; row < Constants.GridSize; row++)
        {
            for (var col = 0; col < Constants.GridSize; col++)
            {
                solver.Grid[row][col] = (row + col) % 2 == 0
                    ? ((row * 3 + col + 1) % 9) + 1
                    : Constants.EmptyCell;
            }
        }
    }

    public static void SetupRowMissingDigit(SudokuSolver solver, int rowIndex, int target)
    {
        var digits = DigitsExcept(target);
        var index = 0;
        for (var col = 0; col < Constants.GridSize; col++)
        {
            if (col == 4)
            {
                solver.Grid[rowIndex][col] = Constants.EmptyCell;
            }
            else
            {
                solver.Grid[rowIndex][col] = digits[index++];
            }
        }
    }

    public static void SetupColumnMissingDigit(SudokuSolver solver, int colIndex, int target)
    {
        var digits = DigitsExcept(target);
        var index = 0;
        for (var row = 0; row < Constants.GridSize; row++)
        {
            if (row == 4)
            {
                solver.Grid[row][colIndex] = Constants.EmptyCell;
            }
            else
            {
                solver.Grid[row][colIndex] = digits[index++];
            }
        }
    }

    public static void SetupRowColumnConstraints(SudokuSolver solver, int count, int rowIndex, int target)
    {
        const int candidateCol = 4;
        var usedRows = new HashSet<int>();
        var blocked = 0;

        for (var col = 0; col < Constants.GridSize && blocked < count; col++)
        {
            if (col == candidateCol)
            {
                continue;
            }

            var row = BlockingRow(col, rowIndex, candidateCol, usedRows);
            solver.Grid[row][col] = target;
            usedRows.Add(row);
            blocked++;
        }
    }

    public static void SetupColumnRowConstraints(SudokuSolver solver, int count, int colIndex, int target)
    {
        const int candidateRow = 4;
        var usedCols = new HashSet<int>();
        var blocked = 0;

        for (var row = 0; row < Constants.GridSize && blocked < count; row++)
        {
            if (row == candidateRow)
            {
                continue;
            }

            var col = BlockingCol(row, candidateRow, colIndex, usedCols);
            solver.Grid[row][col] = target;
            usedCols.Add(col);
            blocked++;
        }
    }

    public static void SetupBlockFourEmpties(SudokuSolver solver)
    {
        solver.Grid[0][0] = 1;
        solver.Grid[1][0] = 2;
        solver.Grid[2][0] = 3;
        solver.Grid[2][1] = 6;
        solver.Grid[2][2] = 7;
    }

    public static void SetupHiddenSingleInBlock(SudokuSolver solver, int target)
    {
        solver.Grid[0][5] = target;
        solver.Grid[3][1] = target;
    }

    public static void SetupDigitInRow(SudokuSolver solver, int rowIndex, int digit) =>
        solver.Grid[rowIndex][5] = digit;

    public static void ClearCell(SudokuSolver solver, int row, int col) =>
        solver.Grid[row][col] = Constants.EmptyCell;

    public static void SetupThreeCandidates(SudokuSolver solver)
    {
        solver.Grid[0][1] = 1;
        solver.Grid[0][2] = 3;
        solver.Grid[0][3] = 4;
        solver.Grid[1][0] = 6;
        solver.Grid[2][0] = 7;
        solver.Grid[4][0] = 9;
    }

    public static void SetupThreeNakedSingles(SudokuSolver solver)
    {
        var solution = new[]
        {
            new[] { 5, 3, 4, 6, 7, 8, 9, 1, 2 },
            new[] { 6, 7, 2, 1, 9, 5, 3, 4, 8 },
            new[] { 1, 9, 8, 3, 4, 2, 5, 6, 7 },
            new[] { 8, 5, 9, 7, 6, 1, 4, 2, 3 },
            new[] { 4, 2, 6, 8, 5, 3, 7, 9, 1 },
            new[] { 7, 1, 3, 9, 2, 4, 8, 5, 6 },
            new[] { 9, 6, 1, 5, 3, 7, 2, 8, 4 },
            new[] { 2, 8, 7, 4, 1, 9, 6, 3, 5 },
            new[] { 3, 4, 5, 2, 8, 6, 1, 7, 9 },
        };

        for (var row = 0; row < Constants.GridSize; row++)
        {
            for (var col = 0; col < Constants.GridSize; col++)
            {
                solver.Grid[row][col] = solution[row][col];
            }
        }

        solver.Grid[0][0] = Constants.EmptyCell;
        solver.Grid[4][4] = Constants.EmptyCell;
        solver.Grid[8][8] = Constants.EmptyCell;
    }

    public static void SetupNamedGridState(SudokuSolver solver, string gridState)
    {
        if (gridState == "has5InSameRow")
        {
            solver.Grid[0][5] = 5;
        }
        else if (gridState == "has3InSameCol")
        {
            solver.Grid[5][0] = 3;
        }
        else if (gridState == "has7InSameBlock")
        {
            solver.Grid[0][0] = 7;
        }
        else if (gridState == "has1InRowAndCol")
        {
            solver.Grid[8][0] = 1;
            solver.Grid[0][8] = 1;
        }
        else if (gridState == "fullyConstrained")
        {
            solver.Grid[3][0] = 8;
        }
    }

    public static void SetupWithDuplicateInRow(SudokuSolver solver, int rowIndex, int value)
    {
        solver.Grid[rowIndex][0] = value;
        solver.Grid[rowIndex][1] = value;
    }

    public static void AddValuesToRow(SudokuSolver solver, int row, int excludeCol, IReadOnlyList<int> values)
    {
        var placed = 0;
        for (var col = 0; col < Constants.GridSize && placed < values.Count; col++)
        {
            if (col != excludeCol && solver.Grid[row][col] == Constants.EmptyCell)
            {
                solver.Grid[row][col] = values[placed++];
            }
        }
    }

    public static void AddValuesToColumn(SudokuSolver solver, int col, int excludeRow, IReadOnlyList<int> values)
    {
        var placed = 0;
        for (var row = 0; row < Constants.GridSize && placed < values.Count; row++)
        {
            if (row != excludeRow && solver.Grid[row][col] == Constants.EmptyCell)
            {
                solver.Grid[row][col] = values[placed++];
            }
        }
    }

    public static void AddValuesToBlock(
        SudokuSolver solver,
        int targetRow,
        int targetCol,
        int excludeRow,
        int excludeCol,
        IReadOnlyList<int> values)
    {
        var blockStartRow = targetRow / Constants.BlockSize * Constants.BlockSize;
        var blockStartCol = targetCol / Constants.BlockSize * Constants.BlockSize;
        var placed = 0;
        for (var row = blockStartRow; row < blockStartRow + Constants.BlockSize && placed < values.Count; row++)
        {
            for (var col = blockStartCol; col < blockStartCol + Constants.BlockSize && placed < values.Count; col++)
            {
                if ((row != excludeRow || col != excludeCol) && solver.Grid[row][col] == Constants.EmptyCell)
                {
                    solver.Grid[row][col] = values[placed++];
                }
            }
        }
    }

    public static IReadOnlyList<SudokuSolver> CreateSolversFromPuzzles(int count, IReadOnlyList<Puzzle> puzzles) =>
        puzzles.Take(count).Select(puzzle => new SudokuSolver(puzzle.Name, puzzle.Grid)).ToArray();

    private static int[] DigitsExcept(int missingDigit) =>
        Enumerable.Range(1, Constants.GridSize).Where(digit => digit != missingDigit).ToArray();

    private static int BlockingRow(int col, int candidateRow, int candidateCol, HashSet<int> usedRows)
    {
        var candidateBlockRow = candidateRow / Constants.BlockSize;
        var candidateBlockCol = candidateCol / Constants.BlockSize;
        var colSharesCandidateBlock = col / Constants.BlockSize == candidateBlockCol;
        var allowedRows = Enumerable.Range(0, Constants.GridSize)
            .Where(row => row != candidateRow && (!colSharesCandidateBlock || row / Constants.BlockSize != candidateBlockRow))
            .ToArray();

        return allowedRows.FirstOrDefault(row => !usedRows.Contains(row), allowedRows[0]);
    }

    private static int BlockingCol(int row, int candidateRow, int candidateCol, HashSet<int> usedCols)
    {
        var candidateBlockRow = candidateRow / Constants.BlockSize;
        var candidateBlockCol = candidateCol / Constants.BlockSize;
        var rowSharesCandidateBlock = row / Constants.BlockSize == candidateBlockRow;
        var allowedCols = Enumerable.Range(0, Constants.GridSize)
            .Where(col => col != candidateCol && (!rowSharesCandidateBlock || col / Constants.BlockSize != candidateBlockCol))
            .ToArray();

        return allowedCols.FirstOrDefault(col => !usedCols.Contains(col), allowedCols[0]);
    }
}
